feat: update UI elements to use `getSelectedInternalAccount` selector

montelaidev opened the pull request:
## **Description**
Currently, MetaMask faces tight coupling between its UI and the keyring-controller. The UI heavily relies on the keyring-controller's state while also amalgamating information from various sources, such as preferences and balances.

However, this approach presents some challenges. The keyring-controller must be aware of the UI's limitations, like the need for instant account list provision to avoid lag. Moreover, it takes on the responsibility of adding metadata to accounts, such as the associated keyring type, required for displaying account details.

To address these issues, the introduction of the accounts-controller comes into play as a new abstraction layer between the UI and the keyring-controller. This separation of responsibilities allows the keyring-controller to focus on two main tasks:

- Routing requests to the appropriate keyring.
- Persisting the state of the keyrings.

This PR replaces the usage of `getSelectedAddress` with `getSelectedInternalAccount`

Depends on: https://github.com/MetaMask/metamask-extension/pull/21554

## **Related issues**

Resolves https://github.com/MetaMask/accounts-planning/issues/135

## **Manual testing steps**

1. Use a hardware wallet to test all the flows

## **Screenshots/Recordings**

No UI/UX changes

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

montelaidev committed:
feat: initial accounts controller

montelaidev committed:
fix: use preview build

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
fix: detect-tokens test

montelaidev committed:
fix: lint

montelaidev committed:
fix: metrics test

montelaidev committed:
fix: remvoe getSelectedAddress from getSnapAndHardwareInfoForMetrics

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
fix: remove getSelectedAddres from args of getSnapAndHardwareInfoForMetrics

montelaidev committed:
fix: lint

montelaidev committed:
fix: e2e tests

montelaidev committed:
fix: update snapshot

montelaidev committed:
fix: add migration

montelaidev committed:
fix: update eth_account test

montelaidev committed:
feat: use release accounts controller 4.0.0 and keyring-controller 8.1.0

montelaidev committed:
fix: restore preference controller to snapkeyring callbacks

montelaidev committed:
Merge branch 'develop' into feat/accounts-controller-2

metamaskbot committed:
Update LavaMoat policies

montelaidev committed:
feat: update eth-snap-keyring to 2.0.0

metamaskbot committed:
Update LavaMoat policies

montelaidev committed:
fix: add getSelectedAccount action to snapAndHardwareMessenger

montelaidev committed:
fix: add getSelectedAccount action to getTransactionMetricsRequest messenger

montelaidev committed:
fix: flaky remove erc1155 e2e test

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
chore: update policy

metamaskbot committed:
Update LavaMoat policies

montelaidev committed:
fix: add waitForSelector for settings tab to add-custom-network

montelaidev committed:
fix: setAccountLabel to setAccountName

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

metamaskbot committed:
Update LavaMoat policies

montelaidev committed:
fix: formatting

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
fix: update to use getAccountByAddress when using listAccount().find() pattern

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
fix: missing action after merge

montelaidev committed:
fix: update comment in migration 103

montelaidev committed:
fix: update 103 migration test

montelaidev committed:
fix: remove unneed resetting of first account

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

montelaidev committed:
fix: detect-tokens test after merge

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into feat/accounts-controller-2

metamaskbot committed:
Update LavaMoat policies

montelaidev committed:
fix: update switchToAccount2 to switchToSnapAccount

montelaidev committed:
fix: remove unused

montelaidev committed:
fix: account name in test-create-snap-account e2e

montelaidev committed:
fix: yarn dedupe

montelaidev committed:
fix: lint

montelaidev committed:
fix: import flow spec

montelaidev committed:
feat: add internal account selectors and update getCurrentKeyring and getAddressBookEntryOrAccountName to use keyring from internal account

montelaidev committed:
fix: revert selectedIdentities to selectedAccounts

montelaidev committed:
fix: update snapshot

montelaidev committed:
fix: add fence for findKeyringForAddress

montelaidev committed:
fix: multichain e2e tests

montelaidev committed:
fix: update setSelectedAccount action to use internalAccount

montelaidev committed:
fix: use getInternalAccounts for confirm-page-container

montelaidev committed:
fix: update hooks to use internal accounts

montelaidev committed:
update selectors/transaction to use getSelectedInternalAccount

montelaidev committed:
fix: update send duck to use InternalAccount

montelaidev committed:
fix: lint

montelaidev committed:
fix: update permissions selector to use internalAccount

montelaidev committed:
fix: update permission-page and signature-request to use InternalAccount

montelaidev committed:
fix: update useAddressDetails to use InternalAccount

montelaidev committed:
fix: update confirm-signature-request, connected accounts, connected-sites, permissions-connect to use InternalAccount

montelaidev committed:
fix: update unconnectedAccount to use internal account

montelaidev committed:
fix: confirm-page-container-content test

montelaidev committed:
fix: add-account spec

montelaidev committed:
fix: update confirm-page and import-tokens-modal to use InternalAccounts

montelaidev committed:
fix: send test, actions test and update confirm-add-suggested-nft snapshot

montelaidev committed:
fix: onboarding test

montelaidev committed:
fix: update ui/index, asset-list, confirm-page-container-content to use InternalAccount

montelaidev committed:
fix: update import-nfts-modal to use InternalAccount

montelaidev committed:
fix: remove unused findKeyringForAddress

montelaidev committed:
fix: update connected-status-indicator, nft-items to use InternalAccounts

montelaidev committed:
fix: revert to use identities

montelaidev committed:
feat: add internal account selectors and update getCurrentKeyring and getAddressBookEntryOrAccountName to use keyring from internal account

montelaidev committed:
fix: update actions and selectors to use getSelectedInternalAccount

montelaidev committed:
fix: readd getSelectedAddress to transactioncontroller

montelaidev committed:
fix: update transaction-list interactive-replacement-token, app-footer, global-menu to use InternalAccount

montelaidev committed:
fix: update selectedAccountWithBalance to default return 0x0 and connected-sites to use internalAccounts

montelaidev committed:
fix: getSelectedInternalAccount test

montelaidev committed:
fix: errors snapshot

montelaidev committed:
fix: errors snapshot

montelaidev committed:
fix: errors e2e test

montelaidev committed:
chore: update yarn.lock

montelaidev committed:
fix: update getSelectedAccount to getSelectedInternalAccount

montelaidev committed:
fix: update metametrics to use get number of internal accounts

montelaidev committed:
fix: metametrics test"

gantunesr committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into fix/update-to-use-getSelectedInternalAccount

gantunesr committed:
chore: rm wrong migration 103 test

gantunesr committed:
chore: update policy

gantunesr committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

gantunesr committed:
chore: delete duplicated code

gantunesr committed:
chore: add error message

gantunesr committed:
chore: revert errors.spec.ts

gantunesr committed:
chore: revert opt in initial bg state

gantunesr committed:
chore: revert opt in init ui state

gantunesr committed:
test: improve unit tests

gantunesr commented:
@metamaskbot update-policies

gantunesr committed:
test: fix internalAccounts mock

metamaskbot committed:
Update LavaMoat policies

gantunesr committed:
test: fix storybook E2E

gantunesr committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

gantunesr committed:
Revert "test: fix storybook E2E"

This reverts commit f3a41df896f4561dde682e275bba2cf3f7edb8ac.

gantunesr committed:
chore: update confirmation components

gantunesr committed:
Revert "chore: update confirmation components"

This reverts commit 1d071cbacd6bd6991a534f65baf915b462bc22f8.

gantunesr committed:
test: fix mock

gantunesr committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

gantunesr committed:
chore: lint

gantunesr committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

gantunesr committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

owencraston commented on ui/components/app/asset-list/asset-list.js:
```diff
@@ -65,7 +65,7 @@ const AssetList = ({ onClickAsset }) => {
   const balance = useSelector(getSelectedAccountCachedBalance);
   const balanceIsLoading = !balance;
-  const selectedAddress = useSelector(getSelectedAddress);
+  const { address: selectedAddress } = useSelector(getSelectedAccount);
```
should this be `getSelectedInternalAccount`?

owencraston commented on ui/ducks/send/send.js:
```diff
@@ -45,7 +45,8 @@ import {
   getSelectedAccount,
-  getSelectedAddress,
+  getSelectedInternalAccount,
+  getSelectedInternalAccountWithBalance,
```
is there no way to derive all the information we need from just one of these methods?

gantunesr replied:
Is that only for the benefit of calling on method instead of two in this file?

owencraston replied:
it will save a re render since these methods are state calls and have side effects. That would be the only reason.

gantunesr replied:
Following that logic then we should only have one method that exposes the entire state, by the looks of it, if the balance is not necessary then the selector should not provide it. This is also not a component, it's a thunk action creator generator

owencraston commented:
The send flow also seems to e stuck when sending money to a snap account.
https://github.com/MetaMask/metamask-extension/assets/22918444/d10f2243-4857-48a2-ae5e-a9461bf91f4d

k-g-j commented:
> The send flow also seems to be stuck when sending money to a snap account. https://github.com/MetaMask/metamask-extension/assets/22918444/d10f2243-4857-48a2-ae5e-a9461bf91f4d

Upon testing these changes with send flow, I can confirm same issues @owencraston  has where it gives a loading yellow spinner perpetually. Also, the list of accounts after clicking send does not reflect the proper snap name (also observed by @owencraston)

<img width="348" alt="Screenshot 2023-12-21 at 12 52 57 PM" src="https://github.com/MetaMask/metamask-extension/assets/91970214/92aed5b3-67d2-47b9-bc7c-c7d869aec06a">

gantunesr committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

gantunesr committed:
fix: account name in send flow

montelaidev committed:
fix: send unapproved transaction from not being found.

montelaidev committed:
fix: lint

montelaidev committed:
fix: previousUserTraits in errors test

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

montelaidev committed:
chore: lint

montelaidev committed:
fix: previousUserTraits in errors test

montelaidev committed:
fix: unconnected-account test

montelaidev committed:
fix:  add accountToConnect prop to connected-accounts-list storybook

montelaidev committed:
fix: Add accounts controller action to snapKeyringBuilder

montelaidev committed:
fix: anchorElement propType error for global menu

montelaidev committed:
fix: update connected-accounts story to use internal account

montelaidev committed:
fix: remove nesting of accountToConnect

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

montelaidev committed:
fix: getAccountToConnectToActiveTab to return internal account

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

montelaidev committed:
chore: fix lint

montelaidev commented:
@metamaskbot update-policies

HowardBraham committed:
Merge branch 'develop' into fix/update-to-use-getSelectedInternalAccount

HowardBraham committed:
trying to fix the E2E failure on "Dapp interactions should connect a second Dapp despite MetaMask being locked"

owencraston commented:
Did a quick qa of the branch and the things look pretty good. I noticed that the account names are incorrect in the send flow which was patched by @gantunesr [in this pr last week](https://github.com/MetaMask/metamask-extension/pull/22482/files).
<img width="441" alt="Screenshot 2024-01-16 at 1 04 57 PM" src="https://github.com/MetaMask/metamask-extension/assets/22918444/53e7c844-c39c-4ecc-9979-ccd818d54141">
![Screenshot 2024-01-16 at 1 00 10 PM](https://github.com/MetaMask/metamask-extension/assets/22918444/3b2b7e71-e7f2-446d-a7f3-72cf72f34a52)

owencraston commented on ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:
```diff
@@ -43,16 +42,8 @@ const UnconnectedAccountAlert = () => {
-  };
-  const selectedAddress = useSelector(getSelectedAddress);
+  const account = useSelector(getSelectedInternalAccount);
+  const { address: selectedAddress } = account;
```
I do not think we should remove the logic for the account names as it results in different names in the account list then in the send flow. A better solution would ideally be to export the account name metadata from this selector.

owencraston replied:
retested with the later commits and this is no longer relevant.

k-g-j commented on ui/components/multichain/app-footer/app-footer.js:
```diff
@@ -55,7 +55,10 @@ export const AppFooter = () => {
+  const selectedAccount = useSelector(getSelectedInternalAccount);
+
+  // During onboarding there is no selected internal account
+  const selectedAddress = selectedAccount?.address;
```
Would `const selectedAddress = useSelector(getSelectedAddress)` ever return null? Would a undefined value have an effect on 
```ts
  const connectedSite = connectedSites[selectedAddress]?.find(
    ({ origin }) => origin === currentTabOrigin,
  );
  ```

montelaidev replied:
It could during onboarding when there aren't any accounts. The connected sites wouldn't be found, but thats ok since we are still setting up the account.

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

owencraston reviewed (dismissed):
tested with an account snap with the send flow and sign data. All the flows seem to work. The selector code changes are nice and minimal and I appreciate the extra tests. 🐐

k-g-j commented on ui/components/multichain/import-tokens-modal/import-tokens-modal.js:
```diff
@@ -459,7 +459,9 @@ export const ImportTokensModal = ({ onClose }) => {
-      case Boolean(identities[standardAddress]):
+      case Boolean(
+        accounts.find((internalAccount) => internalAccount.address === address),
+      ):
```
Why the change here to use the original address and not the standardized version?

montelaidev committed:
fix: change to use standard address

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

montelaidev committed:
chore: update snapshot

montelaidev committed:
fix: update address copy button to use getSelectedInternalAccount

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

montelaidev committed:
chore: bump accounts controller and keyring api

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

HowardBraham committed:
Fix "Snap account signatures" flaky tests (#22614)

### The changes

- Upgraded snap-simple-keyring from 1.0.1 to 1.1.1
- `driver.clickElement` on the new `data-testid` that became available
in 1.1.1
- added one `driver.delay`
- Started using `TEST_SNAPS_SIMPLE_KEYRING_WEBSITE_URL` in more places,
which required moving it to a different file for Storybook usage

montelaidev committed:
fix: backup and utils test

montelaidev committed:
chore: update yarn lock and policy

montelaidev commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

chloeYue committed:
Account list e2e test: split account details e2e tests into multiple files and improve test names (#22635)

## **Description**
This PR dose two things:

- Split the tests in previous `account-details.spec.js` file into
multiple files and placed them in one account-menu folder. Since some
tests weren't actually belong to account details being in the
account-details file, I split them into new test files for pin/unpin and
hide/unhide actions.

- I also enhanced some test names to be more descriptive.

## **Related issues**

Fixes: #22565 

## **Manual testing steps**
Run the e2e test.

## **Screenshots/Recordings**

### **Before**


### **After**

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

PrgrmrHarshShukla committed:
Part of #20163 for file ui\components\app\terms-of-use-popup\terms-of-use-popup.js (#20618)

segun committed:
FEAT: Update what's new and settings copy to communicate additional network coverage on extension (#22618)

## **Description**
- What's new copy should be updated to:
> Steer clear of known scams while still preserving your privacy with
security alerts powered by Blockaid. This feature is available on
Arbitrum, Avalanche, BNB chain, Ethereum Mainnet, Linea, Optimism and
Polygon.
> 
> Always do your own due diligence before approving requests.

- Settings copy should be updated to:
> Privacy preserving - no data is shared with third parties. Available
on Arbitrum, Avalanche, BNB chain, Ethereum Mainnet, Linea, Optimism and
Polygon.

## **Related issues**

Fixes:
[#1695](https://github.com/MetaMask/MetaMask-planning/issues/1695)

Blocked By: #22070 

## **Screenshots/Recordings**

### **Before**


https://github.com/MetaMask/metamask-extension/assets/44811/0e6b6103-efb0-4ed8-94c2-44ac0f281ff8

### **After**


https://github.com/MetaMask/metamask-extension/assets/44811/36740992-0af2-44c5-b62b-0ac522df4d17

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

digiwand committed:
fix: Blockaid What's New os theme -> dark mode (#22649)

## **Description**

Fix Blockaid What's New image when user setting is OS and their OS
setting is dark mode

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/22646
Related to: https://github.com/MetaMask/metamask-extension/pull/22613

## **Manual testing steps**

1. Have a new instance of the wallet to populate the What's New modal
2. Test the UI with different settings and when OS is light and dark
mode


## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

darkwing committed:
UX: Send Flow: Add 'Sent to Contract' warning to new send page (#22551)

## **Description**

This PR implements the "You're sending to a contract" acknowledgement in
the new send flow. This existed in the old send flow.

## **Related issues**

N/A

## **Manual testing steps**

1. Go to the new send page
2. As the recipient, paste: `0x514910771af9ca656af840dff83e8264ecf986ca`
; this is the $LINK token address
3. See the warning, note that you cannot submit the send until
acknowledgement

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

N/A

### **After**

<img width="415" alt="SCR-20240116-mthm"
src="https://github.com/MetaMask/metamask-extension/assets/46655/87955c6d-262b-4c60-82c6-71a295f42e41">


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

montelaidev committed:
feat: check for ledger error codes in the message and display a more … (#21038)

## **Description**

This pr adds more user friendly error messages instead of just error
codes when a user encounters ledger connection issues.

Resolves https://github.com/MetaMask/metamask-extension/issues/17606

## **Manual testing steps**

1. Go to Add hardware wallet
2. On the ledger device, unlock it.
3. On the `Connect a hardware wallet` page, Click on Ledger and then
continue
4. Do not click connect when the webhid connection window pops up.
5. On the ledger device, lock the device
6. On the browser, click on the nano and then continue
7. See the new error


## **Screenshots/Recordings**

### **Before**


![ledger-error-2](https://github.com/MetaMask/metamask-extension/assets/96463427/67d44ae3-6967-415a-b2ae-bfb4e20911fd)


### **After**


![ledger-error](https://github.com/MetaMask/metamask-extension/assets/96463427/896edf82-57b4-4fe1-b04a-75afcfeed149)


## **Related issues**

Resolves https://github.com/MetaMask/metamask-extension/issues/17606

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@consensys.net>

Gudahtt committed:
chore: Upgrade `@metamask/controller-utils` to `v8.0.1` (#22500)

## **Description**

The package `@metamask/controller-utils` has been upgraded to `v8.0.1`.
The breaking changes did not affect any of the imports used by the
extension.

See here for the changelog:
https://github.com/MetaMask/core/blob/main/packages/controller-utils/CHANGELOG.md#801

## **Related issues**

N/A

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: MetaMask Bot <metamaskbot@users.noreply.github.com>
Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>

k-g-j reviewed (dismissed):
Reviewed code and satisfied with the thorough unit tests and resolutions to my initial comments. I tested using both my regular EOA and a snap account for sending and signing flows and did not observe any bugs.

segun committed:
feat: Enable ppom on other chains (#22633)

makoshan committed:
Add imToken Wallet support with EIP-4527 (#21804)

## **Description**
Hello, this is the imToken team. 

imToken is a mobile crypto wallet founded in 2016. It has been operating
safely and steadily for 7 years. Recently, we successfully supported
ERC-4527, and imToken can function as a QR code-based signer now. This
integration seamlessly aligns with Metamask's bidirectional QR account
feature, enhancing the ability to track accounts whose private keys are
stored on external devices.

Therefore, we propose that Metamask expand its supported QR code-based
wallet connection methods and include imToken. This will not only
enhance Metamask's functionality and user coverage, but also provide
users with a wider, more secure and efficient choice of wallets.

We have successfully completed the integration of ERC-4527 and look
forward to discussing in depth the possibility of adding imToken into
Metamask's QR code-based wallet connection methods. Thank you very much
for your attention and feedback on this matter.

## **Related issues**
no
## **Manual testing steps**
1. Please check if the added links to the imToken official website and
tutorials are correct.
2. The process is consistent with the current QR-Based Keystone and
Airgap methods. Please refer to the test video provided below for
reference.

### Video
[Dropbox
link](https://www.dropbox.com/scl/fi/kbfcthouwgqaaueepsslc/imToken_Connect_Metamask_.mp4?rlkey=vz6js9g81oevg2y77vkizfefq&dl=0)
### User Flow

![metamask](https://github.com/MetaMask/metamask-extension/assets/7024451/9280a9dd-de4e-4635-b3db-eb44f3276b6b)

## **Screenshots/Recordings**
### **Before**
<img width="571" alt="image"
src="https://github.com/MetaMask/metamask-extension/assets/7024451/e67a1c52-4c1a-46eb-8140-8df9acc8556d">

### **After**
<img width="346" alt="image"
src="https://github.com/MetaMask/metamask-extension/assets/7024451/3246af50-7187-49a4-b7f5-a84eaf1874f7">

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@gmail.com>
Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@consensys.net>

montelaidev commented:
@SocketSecurity ignore npm/@metamask/keyring-api@3.0.0
@SocketSecurity ignore npm/@metamask/eth-keyring-controller@17.0.1
@SocketSecurity ignore npm/@metamask/keyring-controller@12.2.0
@SocketSecurity ignore npm/@metamask/accounts-controller@10.0.0
@SocketSecurity ignore npm/@metamask/obs-store@9.0.0

Ok bumps in controllers and keyring api

montelaidev committed:
Merge remote-tracking branch 'origin/develop' into fix/update-to-use-getSelectedInternalAccount

montelaidev committed:
fix: update errors snapshot

montelaidev commented:
@SocketSecurity ignore npm/rfdc@1.3.1

The dependency has been verified and confirmed with @FrederikBolding.
